Multioutput regression fix

eddiebergman opened the pull request:
This PR fixes issue #1169 in which the output shape of the predictions with CV models was not as expected and unusable.

# Changes
* `_model_predict()` in `automl.py` now checks for multioutput regression and converts accordingly.
* Changed the logger warnings about size mismatches to assertion errors. This also removes the need for the logger param.
* Updated the code in `_model_predict` to give assertion errors in general as we do not expect use errors to get this far.
* Created tests to catch issue #1169 and any other output shape mismatches with `_model_predict()`
* SUPPORTED_{FEAT/TARGET}_TYPES now generalize to use `spmatrix` instead of listing each kind of sparse matrix.
* Doc and code legibility improvements to `convert_multioutput_multiclass_to_multilabel()` and `_model_predict()`

# Diagnoses of issue
* Mostly described in #1169
* CV models are ensembled together in a `VotingRegressor` which does not support multioutput regression. Normally this is caught in `fit()` but we bypass this as the models are already fitted.
* `VotingRegressor` then treats the models outputs as if they're 2d and averages along the wrong axis, causing the shape mistmatch.

# Future work
* A better method to ensemble together CV models, please see issue #1169 for a more full description of the problem.

mfeurer commented on autosklearn/automl.py:
```diff
@@ -75,38 +75,82 @@
+    model: Any
+        The model to perform predictions with
+
+    X: array-like (n_samples, ...)
```
This can also be sparse, scikit-learn disambiguates between array-like and sparse matrices, see for example [here](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.fit)

mfeurer commented on autosklearn/automl.py:
```diff
@@ -75,38 +75,82 @@
+
+    batchsize: Optional[int] = None
+        If the model supports batch_size predictions then it's possible to pass
+        this in as a parameter.
```
```suggestion
        this in as an argument.
```

mfeurer commented on autosklearn/automl.py:
```diff
@@ -75,38 +75,82 @@
+
+    Returns
+    -------
+    np.ndarray (n_samples, ...)
```
```suggestion
    np.ndarray of shape (n_samples,) or (n_samples, n_outputs)
```

mfeurer commented on autosklearn/automl.py:
```diff
@@ -75,38 +75,82 @@
+
+    assert X_.shape[0] >= 1, f"X must have more than 1 sample but has {X_.shape[0]}"
+
     with warnings.catch_warnings():
```
Is there anything done with the warnings?

eddiebergman replied:
This line was present from before, I assumed there was some reason it was there.
https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings

I think it is better to remove it as it might suppress information we care to know about

mfeurer replied:
IIRC this was there so that the warnings could be swallowed if the user passed `n_jobs>1` to predict(). So I think it would be good to pass warnings to the logger again, as warnings will by default be displayed on the command line.

eddiebergman replied:
This issue was actually caught in the prior version but instead of throwing an error, it was logged as a warning. This is indeed however a separate case as it would only catch warnings and forward them to the logger them as warnings. It does make me fear the sweep-under-the-rug ability of `logger.warning` though.

If there are none or a small amount of occasional warnings, I would be more inclined to let them be displayed on the command line. I would imagine many users don't check the logs for possible warnings that occur, rather they first see them once there's a problem and we ask them to paste the log. Perhaps letting the user know something is being warned against is useful in case that warning leads to them getting output they were not expecting. 

However, if there are many of the same warnings displayed that no actionable work can be done on then yes I agree it's better to just log them.

Are you familiar with what warnings came up? If not I will reintroduce the logger parameter, log them as a warning and then take a note to occasionally browse them myself to see if anything comes up.

eddiebergman replied:
So I made the logger optional and it will log to the logger if passed in. For now it's always passed in.

mfeurer commented on autosklearn/pipeline/implementations/util.py:
```diff
@@ -18,20 +18,51 @@ def softmax(df):
+
+        return multioutput_probas
+
+    else:
```
Maybe we should put the check for `np.ndarray`'s dimension later and do the following to be a bit more restrictive:
```
elif isinstance(probas, np.ndarray):
    if len(probas.shape) > 2:
        raise ValueError('New unsupported sklearn output!')
    else:
        return probas
else:
    raise ValueError('Some error message')
```
what do you think about that?

eddiebergman replied:
I agree with this change :) It's an internal function so we should raise an error if anything unexpected comes though

eddiebergman commented:
I added a contextmanager `warnings_to(Optional[Logger])` that logs to a logger if one is passed, otherwise just continues as if there is no contextmanager. I added tests to this functionality as well so they will need a review
